Framework/Algorithms/src/He3TubeEfficiency.cpp

```diff
@@ -147,60 +146,59 @@ void He3TubeEfficiency::exec() {
- * information. Gets the detector information and uses this to calculate its
- * efficiency
- *  @param spectraIndex :: index of the spectrum to get the efficiency for
- *  @throw invalid_argument if the shape of a detector is isn't a cylinder
- *  aligned along one axis
- *  @throw NotFoundError if the detector or its gas pressure or wall thickness
- *  were not found
- */
+* Corrects a spectra for the detector efficiency calculated from detector
+* information. Gets the detector information and uses this to calculate its
+* efficiency
+*  @param spectraIndex :: index of the spectrum to get the efficiency for
+*  @throw invalid_argument if the shape of a detector is isn't a cylinder
+*  aligned along one axis
+*  @throw NotFoundError if the detector or its gas pressure or wall thickness
+*  were not found
+*/
 void He3TubeEfficiency::correctForEfficiency(std::size_t spectraIndex) {
-  Geometry::IDetector_const_sptr det = this->inputWS->getDetector(spectraIndex);
+  Geometry::IDetector_const_sptr det = m_inputWS->getDetector(spectraIndex);
   if (det->isMonitor() || det->isMasked()) {
     return;
   }
 
   const double exp_constant = this->calculateExponential(spectraIndex, det);
   const double scale = this->getProperty("ScaleFactor");
 
-  Mantid::MantidVec &yout = this->outputWS->dataY(spectraIndex);
-  Mantid::MantidVec &eout = this->outputWS->dataE(spectraIndex);
-  // Need the original values so this is not a reference
-  const Mantid::MantidVec yValues = this->inputWS->readY(spectraIndex);
-  const Mantid::MantidVec eValues = this->inputWS->readE(spectraIndex);
-
-  auto yinItr = yValues.cbegin();
-  auto einItr = eValues.cbegin();
-  auto xItr = this->inputWS->readX(spectraIndex).cbegin();
-  auto youtItr = yout.begin();
-  auto eoutItr = eout.begin();
-
-  for (; youtItr != yout.end(); ++youtItr, ++eoutItr) {
-    const double wavelength = (*xItr + *(xItr + 1)) / 2.0;
-    const double effcorr =
-        this->detectorEfficiency(exp_constant * wavelength, scale);
-    *youtItr = (*yinItr) * effcorr;
-    *eoutItr = (*einItr) * effcorr;
-    ++yinItr;
-    ++einItr;
-    ++xItr;
-  }
+  const auto &yValues = m_inputWS->y(spectraIndex);
+  auto &yOut = m_outputWS->mutableY(spectraIndex);
+
+  const auto wavelength = m_inputWS->points(spectraIndex);
+
+  // todo is it going to die? returning local var by reference
+  std::vector<double> effCorrection(wavelength.size());
+
+  computeEfficiencyCorrection(effCorrection, wavelength, exp_constant, scale);
+
+  std::transform(
+      yValues.cbegin(), yValues.cend(), effCorrection.cbegin(), yOut.begin(),
+      [&](const double y, const double effcorr) { return y * effcorr; });
+
+  const auto &eValues = m_inputWS->e(spectraIndex);
+  auto &eOut = m_outputWS->mutableE(spectraIndex);
+
+  // reset wavelength iterator
+  std::transform(
+      eValues.cbegin(), eValues.cend(), effCorrection.cbegin(), eOut.begin(),
+      [&](const double e, const double effcorr) { return e * effcorr; });
```

Comment:
@LamarMoore and me just discussed this. We concluded that we want to provide `operator*` for multiplying `Histogram` by `std::vector<double>`, then most of this reduces to a single line. Keep in touch with @LamarMoore to see when this is implemented. Probably it will be a follow up to https://github.com/mantidproject/mantid/pull/17193.

Reply:
That PR was merged in September 2016. This can be changed to `eOut = eValues * effCorrection;`